File: app-components/src/test/java/org/synsystems/onlypass/components/cryptography/random/TestSecureByteArrayGenerator.java
package org.synsystems.onlypass.components.cryptography.random;

import org.junit.Before;
import org.junit.Test;

import java.security.SecureRandom;

import static java.util.concurrent.TimeUnit.MILLISECONDS;

@SuppressWarnings("ConstantConditions")
public class TestSecureByteArrayGenerator {
  private SecureByteArrayGenerator generator;

  @Before
  public void setup() {
    generator = new SecureByteArrayGenerator(new SecureRandom());
  }

  @Test(expected = RuntimeException.class)
  public void testConstructor_nullSecureRandom() {
    new SecureByteArrayGenerator(null);
  }

  @Test(expected = RuntimeException.class)
  public void testGenerateWithLength_lengthOfNegativeOne() {
    generator.generateWithLength(-1);
  }

  @Test
  public void testGenerateWithLength_lengthOfZero() {
    generator
        .generateWithLength(0)
        .test()
        .awaitDone(200, MILLISECONDS)
        .assertNoErrors()
        .assertValue(value -> value.length == 0)
        .assertComplete();
  }

  @Test
  public void testGenerateWithLength_lengthOOne() {
    generator
        .generateWithLength(1)
        .test()
        .awaitDone(200, MILLISECONDS)
        .assertNoErrors()
        .assertValue(value -> value.length == 1)
        .assertComplete();
  }
}

File: app-components/src/test/java/org/synsystems/onlypass/components/cryptography/random/TestSecureByteArrayGenerator.kt
package org.synsystems.onlypass.components.cryptography.random

import org.junit.Before
import org.junit.Test

import java.security.SecureRandom

import java.util.concurrent.TimeUnit.MILLISECONDS

class TestSecureByteArrayGenerator {
  
  private lateinit var generator: SecureByteArrayGenerator

  @Before
  fun setup() {
    generator = SecureByteArrayGenerator(SecureRandom())
  }

  @Test(expected = RuntimeException::class)
  fun testGenerateWithLength_lengthOfNegativeOne() {
    generator.generateWithLength(-1)
  }

  @Test
  fun testGenerateWithLength_lengthOfZero() {
    generator
        .generateWithLength(0)
        .test()
        .awaitDone(200, MILLISECONDS)
        .assertNoErrors()
        .assertValue { value -> value.isEmpty() }
        .assertComplete()
  }

  @Test
  fun testGenerateWithLength_lengthOfOne() {
    generator
        .generateWithLength(1)
        .test()
        .awaitDone(200, MILLISECONDS)
        .assertNoErrors()
        .assertValue { value -> value.size == 1 }
        .assertComplete()
  }
}